src/vs/workbench/contrib/mcp/browser/mcpResourceQuickAccess.ts

```diff
@@ -80,13 +113,84 @@ export class McpResourcePickHelper {
 		@INotificationService private readonly _notificationService: INotificationService,
 		@IChatAttachmentResolveService private readonly _chatAttachmentResolveService: IChatAttachmentResolveService
-	) { }
+	) {
+		super();
+	}
+
+	/**
+	 * Navigate to a resource if it's a directory.
+	 * Returns true if the resource is a directory with children (navigation succeeded).
+	 * Returns false if the resource is a leaf file (no navigation).
+	 * When returning true, statefully updates the picker state to display directory contents.
+	 */
+	public async navigate(resource: IMcpResource | IMcpResourceTemplate, server: IMcpServer): Promise<boolean> {
+		try {
+			const uri = await this.toURI(resource);
+			if (!uri) {
+				return false;
+			}
+			const stat = await this._fileService.resolve(uri, { resolveMetadata: false });
+
+			if (this._isDirectoryResource(resource) && (stat.children?.length ?? 0) > 0) {
+				// Save current state to stack before navigating
+				const currentResources = this._resources.get().get(server);
+				if (currentResources) {
+					this.addCurrentMCPQuickPickItemLevel(server, currentResources);
+				}
+
+				// Convert all the children to IMcpResource objects
+				const childResources: IMcpResource[] = stat.children!.map(child => {
+					// const mcpUri = McpResourceURI.fromServer(server.definition, child.resource.toString());
+					return {
+						uri: child.resource,
+						mcpUri: child.resource.path,
+						name: child.name,
+						title: child.name,
+						description: resource.description,
+						mimeType: undefined,
+						sizeInBytes: child.size,
+						icons: McpIcons.fromParsed(undefined)
+					};
+				});
+				this._inDirectory.set({ server, resources: childResources }, undefined);
+				return true;
+			}
+		} catch (error) {
+			// If resolution fails, treat as a leaf (file)
+		}
+		return false;
+	}
+
+	public toAttachment(resource: IMcpResource | IMcpResourceTemplate, server: IMcpServer): Promise<ChatContextPickAttachment> | 'noop' {
+		const noop = 'noop';
+		let attachmentResult: ChatContextPickAttachment = noop;
+
+		const deferred = new DeferredPromise<ChatContextPickAttachment>();
+		if (this._isDirectoryResource(resource)) {
+			//Check if directory
+			this.checkIfDirectoryAndPopulate(resource, server);
+			return noop;
+		}
 
-	public async toAttachment(resource: IMcpResource | IMcpResourceTemplate): Promise<IChatRequestVariableEntry | undefined> {
 		if (isMcpResourceTemplate(resource)) {
-			return this._resourceTemplateToAttachment(resource);
+			this._resourceTemplateToAttachment(resource).then(val => {
+				attachmentResult = (val as IChatRequestVariableEntry) || noop;
+				deferred.complete(attachmentResult);
+			});
 		} else {
-			return this._resourceToAttachment(resource);
+			this._resourceToAttachment(resource).then(val => {
+				attachmentResult = (val as IChatRequestVariableEntry) || noop;
+				deferred.complete(attachmentResult);
+			});
+		}
+		return deferred.p;
```

Comment:
I don't think you need a deferred here. Also the `as` casts are not correct -- the methods can return undefined if the user cancels template URI creation, and that should be bubbled up to the caller.

```suggestion
		if (this._isDirectoryResource(resource)) {
			//Check if directory
			this.checkIfDirectoryAndPopulate(resource, server);
			return noop;
		}

		if (isMcpResourceTemplate(resource)) {
			return this._resourceTemplateToAttachment(resource)
		} else {
			return this._resourceToAttachment(resource);
		}
```

Reply:
Agree, deferred was introduced earlier when checkIfDirectoryAndPopulate was returning a value.No longer needed.